phases/ephemeral/witx/typenames.witx

```diff
@@ -614,24 +612,24 @@
 ;;; The contents of a `subscription` when type is type is
 ;;; `eventtype::fd_read` or `eventtype::fd_write`.
 (typename $subscription_fd_readwrite
-  (struct
+  (record
     ;;; The file descriptor on which to wait for it to become ready for reading or writing.
     (field $fd $fd)
   )
 )
 
 ;;; The contents of a `subscription`.
 (typename $subscription_u
-  (union $eventtype
-    (field $clock $subscription_clock)
-    (field $fd_read $subscription_fd_readwrite)
-    (field $fd_write $subscription_fd_readwrite)
+  (union (@witx tag $eventtype)
+    $subscription_clock
+    $subscription_fd_readwrite
+    $subscription_fd_readwrite
```

Comment:
It's not immediately clear why some `union`s such as `event_u` are migrated to `variant` and others, such as `subscription_u` here, are left as `union`.

Reply:
The desugaring is a bit tricky with the current ABI, but I generally tried to keep things as `union` where possible. By using `@witx tag` the `union`'s types must all mattch up exactly with the variants of the `tag` (which should be an `enum`).

One place this was tripped up though was the `empty` case parsing for unions. That was where one entry of the `union` didn't have any payload. This didn't have any translation to this updated `union` syntax where it must be a set of types that are union'd, so anything using `empty` was switch to the explicit `variant`.

In general it doesn't matter too too much except from a readability perspective since this all desugars to the same thing in the AST, which is to say a tagged `variant`.

Reply:
Is `variant` a superset of the functionality of `union`? If so, I think it's ok to migrate to `variant` over time. To help readers, I do want to aim for a single syntax, but we can finish the migration later.

Reply:
It is, yeah, mostly because `variant` is the actual thing in the AST whereas `union` is just syntax sugar as [seen here ](https://github.com/WebAssembly/interface-types/blob/master/proposals/interface-types/Explainer.md#interface-types).

This will hopefully become less important over time as `@witx tag` is migrated away from perhaps? Otherwise though there could perhaps be more syntax sugar for a `union` where one of the cases is empty.